lib/data_plane/path.c

```diff
@@ -238,85 +221,27 @@ void scion_path_print(const struct scion_path *path)
 	} else if (path->path_type == SCION_PATH_TYPE_SCION) {
 		(void)printf("Hops: [");
-		scion_path_print_interfaces(path->metadata->interfaces);
+		scion_path_print_interfaces(path->metadata->interfaces, path->metadata->interfaces_len);
 		(void)printf("] MTU: %" PRIu32 "\n", path->metadata->mtu);
 	}
 }
 
-uint32_t scion_path_byte_size(struct scion_path *path, bool print_details)
-{
-	if (path == NULL) {
-		return 0;
-	}
-
-	uint32_t size = sizeof(struct scion_path);
-	if (print_details) {
-		(void)printf("		-- scion_path struct itself takes: %" PRIu32 " bytes\n", size);
-	}
-
-	if (path->raw_path != NULL) {
-		uint32_t raw_path_size = sizeof(struct scion_path_raw);
-		raw_path_size += path->raw_path->length;
-		if (print_details) {
-			(void)printf("		-- raw path size (struct + path buffer): %" PRIu32 " bytes\n", raw_path_size);
-			(void)printf("			          path->raw_path->length: %" PRIu16 " bytes\n", path->raw_path->length);
-			(void)printf(
-				"			          sizeof(struct scion_path_raw): %zu bytes\n", sizeof(struct scion_path_raw));
-		}
-		size += raw_path_size;
-	}
-	if (path->metadata != NULL) {
-		uint32_t meta_size = sizeof(struct scion_path_metadata);
-
-		uint32_t intf_size = 0;
-		if (path->metadata->interfaces != NULL) {
-			intf_size += sizeof(struct scion_linked_list);
-			struct scion_linked_list_node *curr = path->metadata->interfaces->first;
-			while (curr) {
-				intf_size += sizeof(struct scion_linked_list_node);
-				if (curr->value != NULL) {
-					intf_size += sizeof(struct scion_path_interface);
-				}
-				curr = curr->next;
-			}
-			meta_size += intf_size;
-		}
-		if (print_details) {
-			(void)printf("		-- total metadata size: %" PRIu32 " bytes\n", meta_size);
-			if (path->metadata->interfaces != NULL) {
-				(void)printf("			of which: %" PRIu32 " interfaces size: %" PRIu32 " bytes\n",
-					path->metadata->interfaces->size, intf_size);
-			} else {
-				(void)printf("			of which: interfaces size: %" PRIu32 " bytes\n", intf_size);
-			}
-
-			(void)printf(
-				"			          sizeof(struct scion_linked_list): %zu bytes\n", sizeof(struct scion_linked_list));
-			(void)printf("			          sizeof(struct scion_linked_list_node): %zu bytes\n",
-				sizeof(struct scion_linked_list_node));
-			(void)printf("			          sizeof(struct scion_path_interface): %zu bytes\n",
-				sizeof(struct scion_path_interface));
-		}
-		size += meta_size;
-	}
-	if (print_details) {
-		(void)printf("	-- Total size: %" PRIu32 " bytes\n", size);
-	}
-	return size;
-}
-
-uint32_t scion_path_get_weight(const struct scion_path *path)
+size_t scion_path_get_hops(const struct scion_path *path)
 {
 	assert(path != NULL);
 
-	return path->weight;
+	if (path->path_type == SCION_PATH_TYPE_EMPTY) {
+		return 0;
+	} else {
+		return (path->metadata->interfaces_len / 2) + 1;
+	}
 }
 
-uint32_t scion_path_get_mtu(const struct scion_path *path)
+struct scion_path_metadata *scion_path_get_metadata(struct scion_path *path)
```

Comment:
Should that maybe return a `const`? And the argument could be `const`?

Reply:
I agree that it would be very nice to make the return type of this function `const`. This would ensure that a user of the library only reads the metadata and does not accidentally modify it.

The argument can also be made `const` but is less usefull.

Reply:
I forgot what the plan is for sharing paths between sockets. Do we copy paths or do we make them immutable? If it is the latter then making the argument const would be useful?
I guess the question is more whether we do it now or later?

Reply:
The idea would be to prevent accidental modification, which would be harmful with concurrent readers. What happens if two sockets try to reverse the path at the same time? It's seems like an unlikely scenario at the moment but I think it would be nice if this could be fully ruled out.

Related to that, I think we discussed that before: JPAN reverses the path in-place when receiving a path. After that it is immutable. This avoids any concurrency issues and checking whether a path is already reversed or not.